src/workspacer.Sound/SoundManager.cs

```diff
@@ -0,0 +1,393 @@
+        public event DeviceEventDelegate DeviceAdded;
+        public event DeviceDataEventDelegate<PROPERTYKEY> DevicePropertyValueChanged;
+        public event DeviceDataEventDelegate<uint> DeviceChanged;
+
+        public event DeviceDataEventDelegate<float> VolumeChanged;
+        public event DeviceDataEventDelegate<bool> MuteChanged;
+
+        public SoundManager()
+        {
+            _mmDeviceEnumerator = Activator.CreateInstance(Type.GetTypeFromCLSID(new Guid(ComCLSIDs.MMDeviceEnumeratorCLSID))) as IMMDeviceEnumerator;
+            if (_mmDeviceEnumerator == null)
+            {
+                throw new Exception("Creating device enumerator failed");
+            }
+
+            _mmDeviceEnumerator.RegisterEndpointNotificationCallback(this);
+        }
+
+        public SoundClient RegisterClientForDefault(DeviceType type, params DeviceRole[] roles)
+        {
+            var deviceId = GetDefaultDeviceId(type, roles);
+            var client = RegisterClient(deviceId, type, roles);
+
+            return client;
+        }
+
+        public SoundClient RegisterClientForDeviceId(string deviceId)
+        {
+            return RegisterClient(deviceId);
+        }
+
+        private SoundClient RegisterClient(string deviceId, DeviceType? type = null, params DeviceRole[] roles)
+        {
+            var client = CreateClient(deviceId, type, roles);
+
+            //Map events
+            client.VolumeChanged += HandleVolumeChanged;
+            client.MuteChanged += HandleMuteChanged;
+
+            _clients.Add(client);
+
+            return client;
+        }
+
+        private void HandleMuteChanged(DeviceInfo device, bool newState)
+        {
+            MuteChanged?.Invoke(device, newState);
+        }
+
+        private void HandleVolumeChanged(DeviceInfo device, float newState)
+        {
+            VolumeChanged?.Invoke(device, newState);
+        }
+
+        public IEnumerable<SoundClient> GetClientsForDeviceId(string deviceId)
+        {
+            return _clients.Where(x => x.DeviceInfo.Id == deviceId);
+        }
+
+        public string GetDefaultPlaybackDeviceId(DeviceRole role = DeviceRole.Multimedia)
+        {
+            return GetDefaultDeviceId(DeviceType.Render, role);
+        }
+
+        public string GetDefaultRecordingDeviceId(DeviceRole role = DeviceRole.Communications)
+        {
+            return GetDefaultDeviceId(DeviceType.Capture, role);
+        }
+
+        public string GetDefaultDeviceId(DeviceType type, params DeviceRole[] roles)
+        {
+            if (_mmDeviceEnumerator == null)
+            {
+                return null;
+            }
+
+            IMMDevice device = null;
+            foreach (var role in roles)
+            {
+                _mmDeviceEnumerator.GetDefaultAudioEndpoint(type.ConvertToEDataFlow(), role.ConvertToERole(), out device);
+                if(device != null)
+                {
+                    break;
```

Comment:
It may be a matter of taste, but simply retuning the device-id inside this if-block would improve code locality, reduce the amount of intermediate variables created, and make it more obvious what happens in the 2 different use-cases (device found, device not found).